### PR TITLE
Feature - API - Home

### PR DESCRIPTION
**Pull request by mvn-khuyendang-dn**

### CODE REVIEW

File: FinalProject/Model/API/Service/HomeService.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                    guard let items = groups.first?["items"] as? JSArray else { return }
+                        recommendVenue = Mapper<RecommendVenue>().mapArray(JSONArray: items)
+                    for venue in recommendVenue {
+                        venue.image = randomImage()
+                    }
+                    completion(.success(recommendVenue))
+
+                } else { return }
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/Model/API/Service/HomeService.swift:59]: else chổ này sao lại return, lỡ nó vào đây thì sao

File: FinalProject/Model/API/Service/HomeService.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                        guard let items = groups.first?["items"] as? JSArray else { return }
+                        recommendVenue = Mapper<RecommendVenue>().mapArray(JSONArray: items)
+                    for venue in recommendVenue {
+                        venue.image = randomImage()
+                    }
+                    completion(.success(recommendVenue))
+
+                } else { return }
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/Model/API/Service/HomeService.swift:83]: tương tự

File: FinalProject/Model/API/Service/HomeService.swift [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                    guard let items = groups.first?["items"] as? JSArray else { return }
+                        recommendVenue = Mapper<RecommendVenue>().mapArray(JSONArray: items)
+                    for venue in recommendVenue {
+                        venue.image = randomImage()
+                    }
+                    completion(.success(recommendVenue))
+
+                } else { return }
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/Model/API/Service/HomeService.swift:108]: tương tự

File: FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
+        return Config.minimumLineSpacingForSection
+    }
+}
+
+// MARK: - UICollectionViewDelegate
+extension NearTableViewCell: UICollectionViewDelegate {
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        if indexPath.row == ((viewModel?.nearVenues.count ?? 0) - 3) {
+            print("bbbb")
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift:72]: ??

File: FinalProject/View/Controller/Home/HomeCell/OpenningCell/OpeningTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+protocol OpeningTableViewCellDelegate: AnyObject  {
+    func cell(_ cell: OpeningTableViewCell, needPerformAction action: OpeningTableViewCell.Action )
+}
+
+final class OpeningTableViewCell: UITableViewCell {
+
+    // MARK: - Enum
+    enum Action {
+        case loadMore
+    }
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/View/Controller/Home/HomeCell/OpenningCell/OpeningTableViewCell.swift:20]: spacing

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                    this.tableView.reloadData()
+                }
+            case .failure(let error):
+                this.alert(msg: error.localizedDescription, handler: nil)
+            }
+        }
+    }
+
+    private func setupDataOpenning(limit: Int = 10) {
+        HUD.show()
```
Comment (mvn-tamnguyen7-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:86]: Sử dụng DispatchGroup

**Review by mvn-tamnguyen7-dn — COMMENTED** [review]
@mvn-khuyendang-dn 
Please check comment

File: FinalProject/Model/API/Service/HomeService.swift [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+import Alamofire
+import ObjectMapper
+
+class HomeService {
+
+    struct HomeParam {
+        static let clientID = "PQXVYOXN4R55FNHKV05EUIF5OR4GZU4F2ITMOGIW3ZA1CKCZ"
+        static let clientSecret = "JEG0HJKBHZNL4ADKBSMRVBFDDFVZWSFCTQ2P2A3UDQXAFAIK"
+        static let version = "20211118"
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:17]: clientID,  clientSecret, version should define in class Api

File: FinalProject/Model/API/Service/HomeService.swift [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    static let addNearParams: JSObject = [
+        "near": "Viet Nam, Da Nang",
+        "limit": "10"
+    ]
+
+    static let addOpenningParams: JSObject = [
+        "near": "Viet Nam, Da Nang",
+        "openNow": "true"
+    ]
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:40]: Create struct Param, for example:

struct Param {
        var query: String
        var ll: String
      
        init(query: String, ll: String) {
            self.query = query
            self.ll = ll
        }

        func toJSON() -> [String: Any] {
            var json: [String: Any] = [:]
            json["query"] = query
            json["ll"] = ll
            ...
            return json
        }
    }

File: FinalProject/Model/API/Service/HomeService.swift [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    class func getRecommendVenues(completion: @escaping Completion<[RecommendVenue]>) {
+        let recommendParams = params.merging(addRecommendParams) { _, _ in }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:43]: Current location of user is dynamic, so please use CLLocationManager to get current location for searching the venues

File: FinalProject/Model/API/Service/HomeService.swift [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        let urlString = Api.Path.baseURL
+        api.request(method: .get, urlString: urlString, parameters: recommendParams ) { result in
+            switch result {
+            case .success(let data):
+                if let data = data as? JSObject,
+                   let response = data["response"] as? JSObject,
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                    guard let items = groups.first?["items"] as? JSArray else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:52]: else must call completion

File: FinalProject/Model/API/Service/HomeService.swift [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
+    }
+
+    class func getNearVenues(completion: @escaping Completion<[RecommendVenue]>) {
+        let nearParams = params.merging(addNearParams) { _, _ in }
+        let urlString = Api.Path.baseURL
+        api.request(method: .get, urlString: urlString, parameters: nearParams ) { result in
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:70]: Similar above

File: FinalProject/Model/API/Service/HomeService.swift [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            switch result {
+            case .success(let data):
+                if let data = data as? JSObject,
+                   let response = data["response"] as? JSObject,
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                        guard let items = groups.first?["items"] as? JSArray else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:77]: else must call completion

File: FinalProject/Model/API/Service/HomeService.swift [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                }
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
+    }
+
+    class func getOpenningVenues(limit: Int, completion: @escaping Completion<[RecommendVenue]>) {
+        let abc = params.merging(addOpenningParams) { (current, _) in current }
+        let openningParams = abc.merging(["limit": "\(limit)"]) { _, _ in }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:94]: Similar above

File: FinalProject/Model/API/Service/HomeService.swift [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        let urlString = Api.Path.baseURL
+        api.request(method: .get, urlString: urlString, parameters: openningParams ) { result in
+            switch result {
+            case .success(let data):
+                if let data = data as? JSObject,
+                   let response = data["response"] as? JSObject,
+                   let groups = response["groups"] as? JSArray {
+                    var recommendVenue: [RecommendVenue] = []
+                    guard let items = groups.first?["items"] as? JSArray else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:103]: else must call completion

File: FinalProject/Model/API/Service/HomeService.swift [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
+    }
+
+    class func randomImage() -> String {
+        let index = Int.random(min: 1, max: 13)
+        return "coffee\(index)"
+    }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/Model/API/Service/HomeService.swift:121]: Define in viewModel

File: FinalProject/View/Base/ImageView.swift [hunk 16]
```diff
@@ -10,3 +10,31 @@ import UIKit
...
+                        completion(image)
+                    } else{
+                        completion(nil)
+                    }
+                }
+            }
+        }
+        task.resume()
+    }
+}
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Base/ImageView.swift:40]: Why extend UIImageview, but not update self.image after download success?

File: FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    // MARK: Life cycle
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        configUINearCollectionView()
+    }
+
+    // MARK: - Private func
+    private func configUINearCollectionView() {
+        let nib = UINib(nibName: "NearCell", bundle: .main)
+        collectionView.register(nib, forCellWithReuseIdentifier: "NearCell")
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift:32]: Please define identifier "NearCell" to use

File: FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+// MARK: - UICollectionViewDataSource
+extension NearTableViewCell: UICollectionViewDataSource {
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        viewModel?.numberOfItemInSection() ?? 0
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "NearCell", for: indexPath) as? NearCell else {
+            return UICollectionViewCell()
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeCell/NearCell/NearTableViewCell.swift:49]: Please define identifier "NearCell" to use

File: FinalProject/View/Controller/Home/HomeCell/OpenningCell/OpeningTableViewCell.swift [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+    // MARK: - Life cycle
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        configUIOpenningCollectionView()
+    }
+
+    // MARK: - Private func
+    private func configUIOpenningCollectionView() {
+        let nib = UINib(nibName: "RecommendCell", bundle: .main)
+        collectionView.register(nib, forCellWithReuseIdentifier: "RecommendCell")
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeCell/OpenningCell/OpeningTableViewCell.swift:42]: Please define identifier "RecommendCell" to use

File: FinalProject/View/Controller/Home/HomeCell/RecommendCell/RecommendTableViewCell.swift [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    // MARK: - Life cycle
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        configUIRecommendCollectionView()
+    }
+
+    // MARK: - Private func
+    private func configUIRecommendCollectionView() {
+        let nib = UINib(nibName: "RecommendCell", bundle: .main)
+        collectionView.register(nib, forCellWithReuseIdentifier: "RecommendCell")
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeCell/RecommendCell/RecommendTableViewCell.swift:32]: Please define identifier "RecommendCell" to use

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+        tableView.register(OpeningTableViewCell.self)
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.showsVerticalScrollIndicator = false
+        tableView.allowsSelection = false
+    }
+
+    private func setupDataRecommend() {
+        HUD.show()
+        guard let viewModel = viewModel else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:55]: What happen when viewModel = nil? HUD will show forever?

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+        viewModel.getRecommendVenues { [weak self]result in
+            HUD.dismiss()
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    this.tableView.reloadRows(at: [IndexPath(row: 0, section: 0)], with: .fade)
+                }
+            case .failure(let error):
+                this.alert(msg: error.localizedDescription, handler: nil)
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:65]: Do need async to main queue before show alert ?

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+            }
+        }
+    }
+
+    private func setupDataNear() {
+        HUD.show()
+        guard let viewModel = viewModel else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:72]: Same above

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+        viewModel.getNearVenues { [weak self] result in
+            HUD.dismiss()
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    this.tableView.reloadRows(at: [IndexPath(row: 1, section: 0)], with: .fade)
+                }
+            case .failure(let error):
+                this.alert(msg: error.localizedDescription, handler: nil)
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:82]: Same above

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+            }
+        }
+    }
+
+    private func setupDataOpenning(limit: Int = 10) {
+        HUD.show()
+        guard let viewModel = viewModel else { return }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:89]: Same above

File: FinalProject/View/Controller/Home/HomeController/HomeViewController.swift [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+            HUD.dismiss()
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    this.tableView.reloadRows(at: [IndexPath(row: 2, section: 0)], with: .fade)
+                }
+            case .failure(let error):
+                this.alert(msg: error.localizedDescription, handler: nil)
+            }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewController.swift:100]: Same above

File: FinalProject/View/Controller/Home/HomeController/HomeViewModel.swift [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// MARK: - HomeViewModel
+extension HomeViewModel {
+    
+    struct Config {
+        static let heightRowOfRecommend: Float = 210
+        static let heightRowOfNear: Float = 200
+        static let heightRowOfOpenning: Float = 970
+    }
```
Comment (mvn-tanhuynh-dn) [inline FinalProject/View/Controller/Home/HomeController/HomeViewModel.swift:114]: Define in enum similar to title

**Review by mvn-tanhuynh-dn — APPROVED** [review]
LGTM!

**Pull request merged**